### PR TITLE
Add configuration option for GOGC, default to 50

### PR DESCRIPTION
**Pull request by SuperQ**
Add the ability to adjust the `GOGC` variable from the Prometheus configuration file.
* Create a new top-level `runtime` section in the config.
* Adjust from the Go default of 100 to 50 to reduce wasted memory.
* Use the `GOGC` env value if no configuraiton is used.

### CODE REVIEW

**Review by bboreham — APPROVED** [review]
I'm ok with this; seems a bit annoying to have to parse the env var ourselves but I guess you looked into that.

The change to the default should be called out in the release notes, since it will provoke a rise in CPU usage and some people might have lots of free memory but little CPU.  And we should write some explanatory text what is happening, since my experience is most people don't know GOGC or the basic mechanism.

**Comment by roidelapluie** [comment]
Use the GOGC env value if no configuraiton is used.

This is the opposite?

File: docs/configuration/configuration.md [hunk 1]
```diff
@@ -121,6 +121,10 @@ global:
   # that will be kept in memory. 0 means no limit.
   [ keep_dropped_targets: <int> | default = 0 ]
 
+  # Configure the Go garbage collector [GOGC parameter](https://tip.golang.org/doc/gc-guide#GOGC)
+  # Lowering this number increases CPU usage.
+  [ gogc: <int> | default = 50 ]
```
Comment (machine424) [inline docs/configuration/configuration.md:126]: because we provide `auto-gomemlimit.ratio` as a CL argument (and gomaxprocs auto setting as arguments as well), I think it's better to have gogc as a arg as well (all config in the same place)

File: cmd/prometheus/main.go [hunk 2]
```diff
@@ -1384,6 +1385,11 @@ func reloadConfig(filename string, expandExternalLabels, enableExemplarStorage b
 		return fmt.Errorf("one or more errors occurred while applying the new configuration (--config.file=%q)", filename)
 	}
 
+	oldGoGC := debug.SetGCPercent(conf.GlobalConfig.GoGC)
```
Comment (machine424) [inline cmd/prometheus/main.go:1388]: we'll need to update how we retrieve it here https://github.com/prometheus/prometheus/blob/efbd6e41c59ec8d6b7a0791c1fb337fdac53b4f2/web/web.go#L721 I think.

**Comment by SuperQ** [comment]
Yes, the Env var parsing is having to reproduce what Go does internally since this isn't exposed. The [source is here](https://github.com/golang/go/blob/eaa7d9ff86b35c72cc35bd7c14b349fa414c392f/src/runtime/mgcpacer.go#L1277-L1286).

**Comment by SuperQ** [comment]
@roidelapluie The logic flow is a bit hard to follow, let me see if I can refactor it a bit to make it more clear.

**Review by machine424 — COMMENTED** [review]
Since we don’t allow automatically setting it like for gomaxprocs or gomemlimit, why do you think that using the GOGC environment variable isn’t adequate/sufficient? I would maintain the default value at 100 (if the environment variable isn’t set), as we can’t guarantee that 50 won’t degrade or cause issues in any setup. We should allow users, who need it, the discretion to select the value that suits them best.

**Comment by SuperQ** [comment]
@machine424 The reason we want to change the default is that several of us have done reasonably extensive production testing that shows that the additional CPU use for Prometheus is pretty minimal.

Especially as we've made major improvements in the overall rate of allocs over the last few years. If I look at metrics history over the last 2 years, the peak alloc bytes/sec/series is 30-40% of what it used to be.

By reducing the Go default from 100 to 50, we see a significant drop in RSS memory use.

File: cmd/prometheus/main.go (hunk 2 above)
Comment (SuperQ) [reply cmd/prometheus/main.go:1388]: Oh, I had an earlier PR that set the env var live to solve that view. I'll put that back in.

File: docs/configuration/configuration.md (hunk 1 above)
Comment (SuperQ) [reply docs/configuration/configuration.md:126]: I would actually like to migrate those flags to the config file. This is mostly my first attempt at adding such a runtime tuneable to the config.

Comment (SuperQ) [reply docs/configuration/configuration.md:126]: I added the `gogc` entry to `global` as there was already a struct for it. But maybe we should have a `runtime` top-level config key? Not sure what other opinions are about that.

File: config/config.go [hunk 3]
```diff
@@ -549,6 +554,31 @@ func (c *GlobalConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
...
+				gc.GoGC = -1
+			} else {
+				i, err := strconv.Atoi(goGCEnv)
+				if err == nil {
+					gc.GoGC = i
+				}
+			}
+		}
+	}
+	// Write the new setting out to the ENV var for runtime API output.
```
Comment (machine424) [inline config/config.go:575]: How about setting this next to `oldGoGC := debug.SetGCPercent` for better visibility and to move it away from this `UnmarshalYAML`?

File: docs/configuration/configuration.md (hunk 1 above)
Comment (machine424) [reply docs/configuration/configuration.md:126]: I see, we can keep it for now I think, and group it with the others when they're migrated.

Comment (SuperQ) [reply docs/configuration/configuration.md:126]: Well, I'd rather avoid moving config keys.

So the question is, do we want to keep it in `global`, add `global.runtime`, or add `runtime` as a top-level key. Maybe some other folks can weigh in on this.

File: config/config.go (hunk 3 above)
Comment (SuperQ) [reply config/config.go:575]: Sure, that also works.

**Comment by machine424** [comment]
> e424 The reason we want to change the default is that several of us have done reasonably extensive production testing that shows that the additional CPU use for Prometheus is pretty minimal.
> 
> Especially as we've made major improvements in the overall rate of allocs over the last few years. If I look at metrics history over the last 2 years, the peak alloc bytes/sec/series is 30-40% of what it used to be.
> 
> By reducing the Go default from 100 to 50, we see a significant drop in RSS memory use.

Thanks for the context and the details, this looks promising.
Ok, let's give it a try, and if we start yelling about it, we can always adjust or rollback ;)

File: docs/configuration/configuration.md (hunk 1 above)
Comment (machine424) [reply docs/configuration/configuration.md:126]: I misread, I thoughts you were talking about the internal structs.
+1 for `runtime` as a top-level key.

Comment (bwplotka) [reply docs/configuration/configuration.md:126]: Slack conv: https://cloud-native.slack.com/archives/C05U6G6T8G7/p1717487320757739

+1 for runtime

Comment (simonpasquier) [reply docs/configuration/configuration.md:126]: weak +1 for `global.runtime`.

`global` already carries fields unrelated to the scrape configs (like `query_log_file`) and the `runtime`  sub-entry makes it easier to reason about.

**Comment by SuperQ** [comment]
I've refactored this to create a new top-level `runtime` section.

**Review by machine424 — APPROVED** [review]
lgtm, thanks.

**Comment by SuperQ** [comment]
Moving the configuration to a separate top level function seems to break the default env var flow. The `UnmarshalYAML()` seems to not be called if there is no `runtime` top level key in the config.

**Comment by SuperQ** [comment]
Ok, I figured out how to get the empty runtime section to work with the ENV var set.

**Comment by machine424** [comment]
Good catch!, I’d ask for a regression test for that case, but I think you’ve had enough haha.
I’ll add some when I have some time ;)
lgtm
